Fix infinite loop when using HasSortableRelations on a model with a self-referencing relationship

mjauvin opened the pull request:
I have model that has a `belongsToMany` relation on itself and somehow, calling the relation on the model (i.e. `$model->$relationName()`) creates an infinite loop.

Using only the model relation definition solves the issue.

mjauvin committed:
only use model relation definition, bo need to instanciate the relation

LukeTowers commented on src/Database/Traits/HasSortableRelations.php:
```diff
@@ -57,19 +57,18 @@ public function initializeHasSortableRelations() : void
-                    $relationType = $this->getRelationType($relationName);
-                    $this->$relationType[$relationName] = $definition;
-                }
+            $realtionType = $this->getRelationType($relationName);
+            if (!in_array($realtionType, ['belongsToMany', 'morphToMany'])) {
```
```suggestion
            $relationType = $this->getRelationType($relationName);
            if (!in_array($relationType, ['belongsToMany', 'morphToMany'])) {
```

LukeTowers committed:
Update src/Database/Traits/HasSortableRelations.php

LukeTowers commented:
> somehow, calling the relation on the model (i.e. $model->$relationName()) creates an infinite loop.

Not a fan of "somehow" issues being solved without understanding exactly how 😜 Can we dig further into why exactly an infinite loop is caused when attempting to access the relation method at this point in time? I would prefer to address that problem (or at least fully understand it) before just giving up and resorting to the relationship definition properties. 

Does this support relations defined via methods instead of the property approach?

mjauvin commented:
> Not a fan of "somehow" issues being solved without understanding exactly how 😜 Can we dig further into why exactly an infinite loop is caused when attempting to access the relation method at this point in time? I would prefer to address that problem (or at least fully understand it) before just giving up and resorting to the relationship definition properties.

It works for all my models except one that has a relation on itself, so we need to look there.

> Does this support relations defined via methods instead of the property approach?

What do you mean "relations defined via methods" ?

mjauvin commented:
@LukeTowers I was able to reproduce which a MUCH simpler model and a trait specifically written to only trigger this behavior:
```php
trait Test
{
    public function initializeTest() : void
    {
        $relation = $this->myrelation();
    }   
}
```
And the model:
```php
class MyModel extends Model
{
    use \StudioAzura\Foundation\Traits\Test;

    public $table = 'mytable';

    public $belongsToMany = [
        'tests' => [
            MyModel::class,
        ],
    ];
}
```

mjauvin commented:
Not sure EXACTLY what is triggering the loop, but the trait initializes itself then for each relation instanciated (which uses the same model), the trait is probably triggered again when the relation is instanciated...

mjauvin commented:
@LukeTowers I did a bit more digging and it turns out Laravel's HasRelationships concern adds a call to `newRelatedInstance()` method which basically create a new instance of the related model... so if the related model is itself (with the HasSortableRelations trait), then that's where the loop gets created.

mjauvin commented:
So in conclusion, I think the fix in this PR is the right way to address the issue.

LukeTowers commented:
@mjauvin Sounds good, is this something we can add a unit test for?

mjauvin commented:
@LukeTowers What exactly do you want to test ?

LukeTowers commented:
Self relations not causing infinite loops :)

mjauvin commented:
How do you test for that? `AssertNoInfiniteLoop()` ?

🤣

mjauvin committed:
Add HasSortableRelans unit test

- make sure models using this trait with a relation to self will not
cause infinite loop.

mjauvin committed:
no need for db stuff here

mjauvin commented:
@bennothommo @LukeTowers Do we really need to use phpstan? I lose so much time just trying to figureout/fix bogus issues it is not really helping...

mjauvin committed:
ignore GeneratorCommand
